src/rtc.rs

```diff
@@ -0,0 +1,448 @@
+                .write(|w| unsafe { w.mt().bit(mt > 0).mu().bits(mu) })
+        });
+
+        Ok(())
+    }
+
+    fn set_year(&mut self, year: u16) -> Result<(), Self::Error> {
+        if (year < 1970) | (year > 2038) {
+            return Err(Error::InvalidInputData);
+        }
+        let (yt, yu) = bcd2_encode(year as u32);
+        self.modify(|regs| regs.dr.write(|w| unsafe { w.yt().bits(yt).yu().bits(yu) }));
+
+        Ok(())
+    }
+
+    /// set date using NaiveDate (ISO 8601 calendar date without timezone)
+    /// WeekDay is set using set_weekday method
+    fn set_date(&mut self, date: &NaiveDate) -> Result<(), Self::Error> {
+        let (yt, yu) = bcd2_encode((date.year() - 1970) as u32);
+        let (mt, mu) = bcd2_encode(date.month());
+        let (dt, du) = bcd2_encode(date.day());
+
+        self.modify(|regs| {
+            regs.dr.write(|w| unsafe {
+                w.dt()
+                    .bits(dt)
+                    .du()
+                    .bits(du)
+                    .mt()
+                    .bit(mt > 0)
+                    .mu()
+                    .bits(mu)
+                    .yt()
+                    .bits(yt)
+                    .yu()
+                    .bits(yu)
+            });
+        });
+
+        Ok(())
+    }
+
+    fn set_datetime(&mut self, date: &NaiveDateTime) -> Result<(), Self::Error> {
+        // Check if unsigned integer affects encoding to bcd
+
+        self.set_24h_fmt();
+        let (yt, yu) = bcd2_encode((date.year() - 1970) as u32);
+        let (mt, mu) = bcd2_encode(date.month());
+        let (dt, du) = bcd2_encode(date.day());
+
+        let (ht, hu) = bcd2_encode(date.hour());
+        let (mnt, mnu) = bcd2_encode(date.minute());
+        let (st, su) = bcd2_encode(date.second());
+
+        self.modify(|regs| {
+            regs.dr.write(|w| unsafe {
+                w.dt()
+                    .bits(dt)
+                    .du()
+                    .bits(du)
+                    .mt()
+                    .bit(mt > 0)
+                    .mu()
+                    .bits(mu)
+                    .yt()
+                    .bits(yt)
+                    .yu()
+                    .bits(yu)
+            });
+        });
+        self.modify(|regs| {
+            regs.tr.write(|w| unsafe {
+                w.ht()
+                    .bits(ht)
+                    .hu()
+                    .bits(hu)
+                    .mnt()
+                    .bits(mnt)
+                    .mnu()
+                    .bits(mnu)
+                    .st()
+                    .bits(st)
+                    .su()
+                    .bits(su)
+                    .pm()
+                    .clear_bit()
+            });
+        });
+
+        Ok(())
+    }
+
+    fn get_seconds(&mut self) -> Result<u8, Self::Error> {
+        let tr = self.regs.tr.read();
+        let seconds = bcd2_decode(tr.st().bits(), tr.su().bits());
+        Ok(seconds as u8)
+    }
```

Comment:
Two general observations for all these `get_*` methods here:
- They never return an error, so there isn't a reason for a `Result` return type, right?
- Why do they take `&mut self` instead of `&self`? These are getters, so I wouldn't expect them to mutate anything

Reply:
This is due to the more general specifications of the RTCC trait

Reply:
Ah yes, I missed that you are implementing a trait here.